src/mscorlib/shared/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -686,6 +686,115 @@ protected EventSource(EventSourceSettings settings, params string[] traits)
+                    byte[] parameterNameBytes = Encoding.ASCII.GetBytes(parameterName);
+                    metadataLength = metadataLength + (uint)parameterNameBytes.Length + 1;
+                }
+
+                byte[] metadata = new byte[metadataLength];
+
+                // Write metadata: evnetID, eventName, keywords, eventVersion, level, param1 type, param1 name...
+                fixed (byte *pMetadata = metadata, pEventNameBytes = eventNameBytes)
+                {
+                    uint offset = 0;
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, eventID);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, pEventNameBytes, (uint)eventNameBytes.Length);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, (byte)0);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, keywords);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, eventVersion);
+                    WriteToBuffer(pMetadata, metadataLength, ref offset, level);
+                    foreach (var parameter in m_eventData[i].Parameters)
+                    {
+                        // Write parameter type.
+                        WriteToBuffer(pMetadata, metadataLength, ref offset, (uint)Type.GetTypeCode(parameter.ParameterType));
+                        
+                        // Write parameter name.
+                        string parameterName = parameter.Name;
+                        byte[] parameterNameBytes = Encoding.ASCII.GetBytes(parameterName);
+                        fixed (byte *pParameterNameBytes = parameterNameBytes)
+                        {
+                            WriteToBuffer(pMetadata, metadataLength, ref offset, pParameterNameBytes, (uint)parameterNameBytes.Length);
+                        }
+
+                        // Write terminated 0.
+                        WriteToBuffer(pMetadata, metadataLength, ref offset, (byte)0);
+                    }
+
+                    IntPtr eventHandle = m_provider.m_eventProvider.DefineEventHandle(eventID, eventName, keywords, eventVersion, level, pMetadata, metadataLength);
+                    m_eventData[i].EventHandle = eventHandle; 
+                }
+            }
+        }
+
+        // Copy src to buffer and modify the offset.
+        // Note: We know the buffer size ahead of time to make sure no buffer overflow.
+        private unsafe void WriteToBuffer(byte *buffer, uint bufferLength, ref uint offset, byte *src, uint srcLength)
+        {
+            Debug.Assert(bufferLength >= (offset + srcLength));
+            for (int i = 0; i < srcLength; i++)
+            {
+                *(byte *)(buffer + offset + i) = *(byte *)(src + i);
+            }
+            offset += srcLength;
+        }
+
+        // Copy byte value to buffer.
+        private unsafe void WriteToBuffer(byte *buffer, uint bufferLength, ref uint offset, byte value)
```

Comment:
To add the terminated 0 when copying eventName and parameterName. I guess this is not needed if I use unicode string instead of ascii.